### ISSUE TITLE
Panic message not printed correctly

### ISSUE THREAD

**Issue by tiziano88** [project-oak/oak#483]
The error currently shows as

```log
2020-01-16 17:03:46,825 ERROR [oak] panic occurred in file 'abitest/tests/src/tests.rs' at line 76: <UNKNOWN MESSAGE>
```

To reproduce, add `assert_eq!(1, 2)` to some function under test, e.g.:

https://github.com/project-oak/oak/blob/f317c1c16d368d937a1b8992aed0affa03d2dd8a/examples/abitest/tests/src/tests.rs#L55-L88

and then run the tests locally with `cargo test`.

Note that `panic!("xxx")` seems to show up fine instead, not sure if `assert_eq` is using a different mechanism behind the scenes.

I also noticed similar errors a few times when running compiled code on the Oak Runtime, but I don't have reproducibility steps for that yet.

@conradgrobler could you take a look? @daviddrysdale may have some thoughts about it too

**Comment by daviddrysdale** [comment]
The `<UNKNOWN MESSAGE>` indicates that the [panic hook](https://doc.rust-lang.org/std/panic/fn.set_hook.html) that we register in [`oak::set_panic_hook`](https://project-oak.github.io/oak/sdk/oak/fn.set_panic_hook.html) has been [unable to retrieve](https://github.com/project-oak/oak/blob/master/sdk/rust/oak/src/lib.rs#L360) the payload from the [`PanicInfo`](https://doc.rust-lang.org/std/panic/struct.PanicInfo.html).  Dunno (yet) why that should happen for a `panic!` generated from the [`assert_eq!` macro](https://doc.rust-lang.org/std/macro.assert_eq.html) but not from an unadorned `panic!`.

**Comment by daviddrysdale** [comment]
The panic details are also lost if a plain `panic!` has arguments (e.g. `panic!("some {:?} thing", 1);`).  Maybe it's a [Rust bug/limitation](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=41db5e9086aadb5f28e0b6fe142d62c5)?

**Comment by daviddrysdale** [comment]
https://github.com/rust-lang/rust/issues/66745 looks potentially relevant.

**Comment by conradgrobler** [comment]
Changing the custom panic hook to handle the downcasting in a way that is more similar to the default hook should resolve it.

https://github.com/rust-lang/rust/blob/7d761fe0462ba0f671a237d0bb35e3579b8ba0e8/src/libstd/panicking.rs#L171-L177

**Comment by conradgrobler** [comment]
Resolved by #485
